File: frontend-web/src/app/services/inventario.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class InventarioService {
  private http = inject(HttpClient);
  private authService = inject(AuthService);
  private apiUrl = `${environment.apiUrl}/inventario`;

  private get headers(): HttpHeaders {
    return new HttpHeaders({
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.authService.token}`
    });
  }

  getStockPorVariante(varianteId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/stock/${varianteId}`);
  }

  getInventarioSucursal(sucursalId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/sucursal/${sucursalId}`, { headers: this.headers });
  }

  registrarMovimiento(data: {
    variante_id: number;
    sucursal_id: number;
    tipo_movimiento: string;
    cantidad: number;
    observaciones?: string;
    sucursal_destino_id?: number;
  }): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/movimiento`, data, { headers: this.headers });
  }

  getKardex(sucursalId?: number, varianteId?: number): Observable<any[]> {
    let params = new HttpParams();
    if (sucursalId) params = params.set('sucursal_id', sucursalId.toString());
    if (varianteId) params = params.set('variante_id', varianteId.toString());
    return this.http.get<any[]>(`${this.apiUrl}/movimientos`, { headers: this.headers, params });
  }

  ajustarStock(data: {
    variante_id: number;
    sucursal_id: number;
    nueva_cantidad: number;
    observaciones?: string;
  }): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/ajustar`, data, { headers: this.headers });
  }
}

File: frontend_mobile/lib/services/inventario_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../constants.dart';
import 'package:shared_preferences/shared_preferences.dart';

class InventarioService {
  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('jwt_token');
  }

  Future<Map<String, String>> _getHeaders() async {
    final token = await _getToken();
    return {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  Future<List<dynamic>> getSucursales() async {
    try {
      final response = await http.get(Uri.parse('${Constants.apiUrl}/sucursales/'));
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return [];
    } catch (_) { return []; }
  }

  Future<Map<String, dynamic>?> getInventarioSucursal(int sucursalId) async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(
        Uri.parse('${Constants.apiUrl}/inventario/sucursal/$sucursalId'),
        headers: headers
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }

  Future<List<dynamic>> getMovimientos(int sucursalId) async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(
        Uri.parse('${Constants.apiUrl}/inventario/movimientos?sucursal_id=$sucursalId'),
        headers: headers
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return [];
    } catch (_) { return []; }
  }

  Future<bool> registrarMovimiento({
    required int varianteId,
    required int sucursalId,
    required String tipoMovimiento,
    required int cantidad,
    String? observaciones,
  }) async {
    try {
      final headers = await _getHeaders();
      final response = await http.post(
        Uri.parse('${Constants.apiUrl}/inventario/movimiento'),
        headers: headers,
        body: jsonEncode({
          'variante_id': varianteId,
          'sucursal_id': sucursalId,
          'tipo_movimiento': tipoMovimiento,
          'cantidad': cantidad,
          if (observaciones != null) 'observaciones': observaciones,
        })
      );
      return response.statusCode == 201 || response.statusCode == 200;
    } catch (_) {
      return false;
    }
  }

  // Helper to fetch variants for the dropdown
  Future<List<dynamic>> searchVariantes() async {
    try {
      final response = await http.get(Uri.parse('${Constants.apiUrl}/productos/'));
      if (response.statusCode == 200) {
        final List<dynamic> productos = jsonDecode(response.body);
        List<dynamic> allVariantes = [];
        for (var p in productos) {
          if (p['variantes'] != null) {
            for (var v in p['variantes']) {
              v['producto_nombre'] = p['nombre']; // inject to display
              allVariantes.add(v);
            }
          }
        }
        return allVariantes;
      }
      return [];
    } catch (_) { return []; }
  }
}
